### PR TITLE
Fix issue 1351

### LINKED ISSUE #1351: Segfault running on `AerSimulator` with empty circuit
### Informations

- **Qiskit Aer version**: 0.9.0
- **Qiskit version**: 0.30.0
- **Python version**: 3.8.10
- **Operating system**: Linux (Ubuntu 20.04)

### What is the current behavior?

Attempting to run a circuit containing no gates on an `AerSimulator` causes a segmentation fault.

### Steps to reproduce the problem

```python
from qiskit.providers.aer import AerSimulator
from qiskit import IBMQ, QuantumCircuit

c = QuantumCircuit()
IBMQ.load_account()
provider = IBMQ.get_provider(hub="ibm-q", group="open", project="main")
b = provider.get_backend("ibmq_santiago")
aersim = AerSimulator.from_backend(b)
job = aersim.run(c, n_shots=1)
```

### What is the expected behavior?

`AerSimulator.run()` should return.

### Suggested solutions

?

### PR DESCRIPTION
**Pull request by chriseclectic**
### Summary

Fixes #1351

### Details and comments

When loading a circuit in C++ the size of an empty circuit was being set to 1 and being populated with a default initialized Op. This caused a crash when this circuit was passed to the noise sampler. This is fixed by correctly initializing an empty to be size 0.

### CODE REVIEW

File: src/framework/circuit.hpp [hunk 1]
```diff
@@ -252,7 +252,8 @@ void Circuit::add_op_metadata(const Op& op) {
 void Circuit::set_params(bool truncation) {
   // Clear current circuit metadata  
   reset_metadata();
-  
+  if (ops.empty()) return;
+
```
Comment (jakelishman) [inline src/framework/circuit.hpp:256]: This early return means that the check a few lines down `if (!ops.empty())` is now guaranteed to be true.

File: src/noise/noise_model.hpp [hunk 2]
```diff
@@ -260,6 +260,10 @@ NoiseModel::param_gate_table_ = {
 Circuit NoiseModel::sample_noise(const Circuit &circ,
                                  RngEngine &rng,
                                  const Method method) const {
+  // Check edge case of empty circuit
+  if (circ.ops.empty()) {
+    return circ;
+  }
```
Comment (jakelishman) [inline src/noise/noise_model.hpp:266]: If `Circuit::set_params` is doing its job correctly, then this shouldn't be necessary, right?  The auto loop over `circ.ops` further down should just operate zero times.

**Review by jakelishman — COMMENTED** [review]
The code change as written does appear to correctly handle the case of empty ops in the noise model, but solving it purely by early return in `Circuit::set_params` seems to suggest that there's some logic error in the rest of that function, I think - by the time it hits `ops.resize(op_idx)`, `op_idx` should still be 0.

I think this happens in line 383 of `Circuit::set_params` - `has_measure` will certainly be `false`, so `head_end` will get set to 1, causing the next loop to be entered once, and `op_idx` to be incremented, even though there weren't any ops.  So then at the end, the vector gets expanded to be size 1, even though I think the expansion is only meant to be a truncation.

I'm slightly worried that there may be more off-by-one errors lurking in there, because I can't convince myself that the current off-by-one is absolutely specific to an empty `ops`.

*edit*: also, early return from `set_params` feels like a potential future bug in itself - are we sure that all the parameters that need to be set are always going to be before that return check?  Or can we absolutely rely on the defaults being correct?  Neither seems obviously safe to me - it feels better if the looping logic in the rest of the function correctly handles empty lists.

File: src/noise/noise_model.hpp (hunk 2 above)
Comment (chriseclectic) [reply src/noise/noise_model.hpp:266]: I pushed another fix for this and tested it worked for the previous segfault error by commenting out the early return statement.

**Review by jakelishman — APPROVED** [review]
This looks much better to me.  I think the correction to the resize amount should now mean that the early returns from both `Circuit::set_params` and from `NoiseModel::sample_noise` are unnecessary now, and only serve as speed optimisations.  For clarity, it may not hurt to remove them since they're superfluous, but are in functions that have some side-effects that aren't (immediately) obviously nulled if the ops are empty.  I'm also not too bothered if you want to keep them, though.

I don't know what frameworks we have in place for testing the C++ code, but if we've got them, it's probably worth sticking in a regression test that empty `ops` remain size 0 after `Circuit::set_params` for the future.

**Pull request merged**